libraries/AC_AutoTune/AC_AutoTune_Multi.cpp

```diff
@@ -0,0 +1,673 @@
+    case PITCH:
+        updating_angle_p_up(tune_pitch_sp, AUTOTUNE_SP_MAX, AUTOTUNE_SP_STEP, target_angle, test_angle_max, test_rate_min, test_rate_max);
+        break;
+    case YAW:
+        updating_angle_p_up(tune_yaw_sp, AUTOTUNE_SP_MAX, AUTOTUNE_SP_STEP, target_angle, test_angle_max, test_rate_min, test_rate_max);
+        break;
+    }
+}
+
+// update gains for the angle p down tune type
+void AC_AutoTune_Multi::updating_angle_p_down_all(AxisType test_axis)
+{
+    switch (test_axis) {
+    case ROLL:
+        updating_angle_p_down(tune_roll_sp, AUTOTUNE_SP_MIN, AUTOTUNE_SP_STEP, target_angle, test_angle_max, test_rate_min, test_rate_max);
+        break;
+    case PITCH:
+        updating_angle_p_down(tune_pitch_sp, AUTOTUNE_SP_MIN, AUTOTUNE_SP_STEP, target_angle, test_angle_max, test_rate_min, test_rate_max);
+        break;
+    case YAW:
+        updating_angle_p_down(tune_yaw_sp, AUTOTUNE_SP_MIN, AUTOTUNE_SP_STEP, target_angle, test_angle_max, test_rate_min, test_rate_max);
+        break;
+    }
+}
+
+// updating_rate_d_up - increase D and adjust P to optimize the D term for a little bounce back
+// optimize D term while keeping the maximum just below the target by adjusting P
+void AC_AutoTune_Multi::updating_rate_d_up(float &tune_d, float tune_d_min, float tune_d_max, float tune_d_step_ratio, float &tune_p, float tune_p_min, float tune_p_max, float tune_p_step_ratio, float rate_target, float meas_rate_min, float meas_rate_max)
+{
+    if (meas_rate_max > rate_target) {
+        // if maximum measurement was higher than target
+        // reduce P gain (which should reduce maximum)
+        tune_p -= tune_p*tune_p_step_ratio;
+        if (tune_p < tune_p_min) {
+            // P gain is at minimum so start reducing D
+            tune_p = tune_p_min;
+            tune_d -= tune_d*tune_d_step_ratio;
+            if (tune_d <= tune_d_min) {
+                // We have reached minimum D gain so stop tuning
+                tune_d = tune_d_min;
+                counter = AUTOTUNE_SUCCESS_COUNT;
+                AP::logger().Write_Event(LogEvent::AUTOTUNE_REACHED_LIMIT);
+            }
+        }
+    } else if ((meas_rate_max < rate_target*(1.0f-AUTOTUNE_D_UP_DOWN_MARGIN)) && (tune_p <= tune_p_max)) {
+        // we have not achieved a high enough maximum to get a good measurement of bounce back.
+        // increase P gain (which should increase maximum)
+        tune_p += tune_p*tune_p_step_ratio;
+        if (tune_p >= tune_p_max) {
+            tune_p = tune_p_max;
+            AP::logger().Write_Event(LogEvent::AUTOTUNE_REACHED_LIMIT);
+        }
+    } else {
+        // we have a good measurement of bounce back
+        if (meas_rate_max-meas_rate_min > meas_rate_max*aggressiveness) {
+            // ignore the next result unless it is the same as this one
+            ignore_next = true;
+            // bounce back is bigger than our threshold so increment the success counter
+            counter++;
+        } else {
```

Comment:
nitpick: I think this line and the line below can be combined into a single "else if" and "ignore_next == false" could be replaced with "!ignore_next".

Reply:
it was existing code.